### ISSUE TITLE
Support for defer/stream

### ISSUE THREAD

**Issue by maraisr** [apollographql/apollo-client#7691]
Raising this issue with goal of gathering feedback about Apollo "re-adopting" the defer/stream capability. There has been huge momentum in recent months around this. [graphiql now supports](https://github.com/graphql/graphiql/pull/1770) defer/stream out of the box and houses a sample implementation for this. We are also seeing leaps being made over in the spec https://github.com/graphql/graphql-spec/pull/742 So its only a matter of time before this will become an expected feature.

I know that this is scheduled for 3.5; but wondering If this is something that I can help contribute back across the line? I gave this is a fair crack over at [`apollo-link-multipart`](https://github.com/maraisr/apollo-link-multipart) using [`meros`](https://github.com/maraisr/meros) so have had some experience in adding it. But believe adding this into the core _would_ enable greater adoption.

_Links~_
- grpahiql support: https://github.com/graphql/graphiql/pull/1770
- https://github.com/graphql/graphql-spec/blob/main/rfcs/DeferStream.md
- https://github.com/graphql/graphql-over-http/blob/main/rfcs/IncrementalDelivery.md
- Sample server implementation: https://github.com/contrawork/graphql-helix
- transport handler: https://github.com/maraisr/meros

**Comment by maraisr** [comment]
@acao thought you'd be interested here

**Comment by benjamn** [comment]
@maraisr Thanks for the nudge! At a very high level, I think the link chain is the right place to handle the multipart HTTP requests, so your past work may be largely transferrable/adaptable, either by integrating this functionality into the default `HttpLink` class, or by having a separate `@apollo/client/link/multipart` link package (let me know if you have any thoughts/preferences about that). On the Apollo Client side, we need to make sure all the right things happen when a link delivers multiple results over time, but I'm optimistic that functionality is already mostly in place.

**Comment by maraisr** [comment]
Personally I feel this probably belongs in the `HttpLink`, but really can go in either place. The defer/stream directives are optionally implemented by the server, which means if a server exposes that we can go with;

1. Apollo will "be ready" in that all HttpLink's will support that transport.
2. If a server supports it, it'll be opt in client support. 
    - My only gripe with this one is that IDE tooling and what-not will give the user the illusion they can use those directives without potentially knowing there also needs to be transport changes.

So with that; my vote is in `HttpLink`—`meros` is transparent to this, so for clients w/o defer/stream support they'd know no difference.

**Comment by acao** [comment]
agreeing with @maraisr here, similar situation with urql

**Comment by maraisr** [comment]
I have embarked on this journey! 🎉 but do need some clarification.

In order for current link chain apis to stay consistent, how do we suppose incremental payloads get observed? Thinking here, "we build up"—in that initial response builds out the `data, errors` object, `.next`'s that, second payload, grabs first, adds to it, then `.next` again so on and so forth.

eg:

for the query:
```graphql
{
    firstName
    ... @defer {
        lastName
    }
}
```

```
=> { data: {firstName: "bob"} }
=> { data: {firstName: "bob", lastName: "the builder"} }
```

```graphql
query MultiDefer {
    user {
        firstName
        ... @defer(label: "lastName") {
            lastName
        }
        contact {
            ... @defer(label: "contact_phoneNumber") {
                 phoneNumber
            }
        }
    }
}
```

```
=> { data: {firstName: "bob", contact: {}} }
=> { data: {firstName: "bob", contact: {}, lastName: "the builder"} }
=> { data: {firstName: "bob", contact: {phoneNumber: 123}, lastName: "the builder"} }
```

Secondly; Where do we think this ends? I could head down the rabbit hole with the `labels`. IncrementalDelivery spec defines an label arg that can be used to identify the payload that came in from others. Granted im not 100% aware of the Apollo api around fragments. But with Relay a component can `useFragment` and some other parent component does the `useQuery`—so just checking if there _could_ a use-case here to use the label to for instance notify all fragments about changes, and not others.

eg with the `MultiDefer` query above:

Seeing as defer's are raced on server side, the phoneNumber _could_ come back before/after the lastName. So in terms of client updates if one component only cares about the phone number the lastName component gets a useless update. 

Now as far as I understand `useQuery` (Apollo) this is a complete document, and fragments are shared just through sharing, but not so much for masking. So this concept may not be useful.

Thirdly; In terms of the `useQuery` api, the result "loading: boolean" would this be `true` till the last incremental payload arrives? Also in the polling stuff, how would this work? Do we kick off a new poll interval once the last payload arrives, which _could_ then potentially not marry up to interval defined?

Further this, I'm almost tempted introduce a `useAsyncQuery` (name??) that does this in a sane way such as introduce a result variable `hasMore: boolean` to indicate if all payloads have arrived, and not support polling or support it within caveats around timings etc... And then for `useQuery` it'll effectively then `await` for the last payload before yielding.

Ill most likely have more questions, so would appreciate some time with ya'll in terms of direction or to spit ball ideas on how to bring and fully leverage this within this client—and once again, happy to spearhead the implementation (progress: https://github.com/maraisr/apollo-client/tree/incremental-delivery). 

cc @kkemple @abernix @acao

**Comment by benjamn** [comment]
@maraisr Sorry for leaving you hanging for so long here!

As long as the incoming data payloads get written into the `InMemoryCache`, all queries and fragments that are watching the cache will be notified automatically of the new data. This could either mean writing the entire result multiple times, with incremental patches applied, or somehow writing only the data that have changed. The latter approach might be more efficient, but trickier. I would prefer for it to work without relying on the `label`, but I see how `label` could help.

I think we will end up introducing a new `NetworkStatus.partial` enum value to indicate that a query result is still missing some deferred parts. The [`NetworkStatus`](https://github.com/apollographql/apollo-client/blob/main/src/core/networkStatus.ts) concept is something that exists in `ApolloClient` outside of `ApolloLink`, so I don't think we can terminate everything within `ApolloLink`.

While I hope that answers at least some of your questions, and I very much appreciate you sharing your branch as a proof of concept, I have to warn you that we (the core team) will probably have an easier time navigating the internals of Apollo Client than you (our fault, not yours), so it might make sense to hold off on a full PR, to protect your personal time. We will look for a ways to use `meros` (thanks for that), and I will definitely be reaching out to you for your review, when we have something ready for review/testing.

Scheduling-wise, this is a priority for [the v3.5 release](https://github.com/apollographql/apollo-client/blob/main/ROADMAP.md#35) (I got it moved up from v3.6), so I hope you won't be waiting long. Thanks again for going above and beyond to get the ball rolling on this project.

**Comment by hwillson** [comment]
Let's re-open this for tracking (we're aiming to have this in `@apollo/client@3.5`).

**Comment by hwillson** [comment]
Linking to https://github.com/apollographql/apollo-feature-requests/issues/3

**Comment by benjie** [comment]
@maraisr If you can make it, add it to Thursday’s WG: https://github.com/graphql/graphql-wg/blob/main/agendas/2021-05-13.md

(Feel free to reach out to me on Slack/Discord/Twitter if you need guidance; I can also help via the WG agenda PR.)

**Comment by hwillson** [comment]
@maraisr are you still interested in talking about this on tomorrow's working group call? @brainkim and myself are planning on attending, and can submit a PR to have this added to the agenda.

**Comment by maraisr** [comment]
I think let me gather my thoughts a little more, lets get this added to next time. Feel like these sorts of meetings are not for time wasters, so want to be more prepared. @hwillson

**Comment by brainkim** [comment]
@maraisr I don’t think I’m blocked on anything as of yet. I’m currently working off the behavior of graphql-helix, but these same issues will likely play a role in defer/stream support for Apollo’s server as well.

> Feel like these sorts of meetings are not for time wasters

I think I’ve done enough research on this that we can do the meeting tomorrow; let me know if you think otherwise.

**Comment by maraisr** [comment]
Are you going to be attending the WG meet tomorrow around this issue? Perhaps I could be a fly on the wall to chime in as its being discussed, if youre going to lead the discussion? @brainkim

**Comment by brainkim** [comment]
@maraisr Seems like it’s a bit too short notice for the relevant actors to get together for the next meeting. I prefer async communication to be honest.

**Comment by avisra** [comment]
Is this ticket still being worked on?

**Comment by maraisr** [comment]
No, I was told to stand down.

**Comment by Amnesthesia** [comment]
Does this mean `@defer` and `@stream` are straight up not supported in Apollo today?

**Comment by benjie** [comment]
They're not yet a part of the GraphQL Draft Spec, we're still iterating on the payload formats/etc - if you're interested, you can read notes from last night's working group here: https://github.com/graphql/graphql-wg/blob/main/notes/2022/2022-05-05.md#deferstream-asynchronous-iterators-of-iterables-versus-of-items-continuation-of-last-wg-meeting-15m-yaacov
